tests/test_messages_queue.cc

```diff
@@ -0,0 +1,482 @@
+  generator(kTestMessage, limit);
+  worker.join();
+
+  std::vector<std::ofstream::pos_type> file_sizes;
+  FinishTask finish_task;
+  q.ProcessArchivedFiles([&file_sizes](bool is_file, const std::string & full_file_path) {
+    TEST_EQUAL(true, is_file);
+    file_sizes.push_back(FileManager::ReadFileAsString(full_file_path).size());
+    return true;
+  }, std::bind(&FinishedCallback, std::placeholders::_1, std::ref(finish_task)));
+  TEST_EQUAL(ProcessingResult::EProcessedSuccessfully, finish_task.get());
+  TEST_EQUAL(size_t(2), file_sizes.size());
+  TEST_EQUAL(size, file_sizes[0] + file_sizes[1]);
+  if (file_sizes[0] < q.kMaxFileSizeInBytes) {
+    TEST_EQUAL(true, file_sizes[1] > q.kMaxFileSizeInBytes);
+  } else {
+    TEST_EQUAL(true, file_sizes[0] > q.kMaxFileSizeInBytes);
+  }
+}
+
+void Test_MessagesQueue_HighLoadAndIntegrity() {
+  // TODO(AlexZ): This test can be improved by generating really a lot of data
+  // so many archives will be created. But it will make everything much more
+  // complex now.
+  const std::string tmpdir = FileManager::GetDirectoryFromFilePath(GenerateTemporaryFileName());
+  CleanUpQueueFiles(tmpdir);
+  ScopedRemoveFile remover(tmpdir + alohalytics::kCurrentFileName);
+  MessagesQueue q;
+  const int kMaxThreads = 300;
+  std::mt19937 gen(std::mt19937::default_seed);
+  std::uniform_int_distribution<> dis(1, std::numeric_limits<char>::max());
+  auto const generator = [&q](char c) { q.PushMessage(std::string(static_cast<size_t>(c), c)); };
+  std::vector<std::thread> threads;
+  size_t total_size = 0;
+  for (int i = 0; i < kMaxThreads; ++i) {
+    char c = dis(gen);
+    total_size += static_cast<size_t>(c);
+    if (i == kMaxThreads / 2) {
+      // At first, messages go into the in-memory queue. Then we initialize
+      // files storage.
+      q.SetStorageDirectory(tmpdir);
+    }
+    std::thread worker([&generator, c]() { generator(c); });
+    threads.push_back(std::move(worker));
+  }
+  TEST_EQUAL(true, total_size > 0);
+  for (auto & thread : threads) {
+    thread.join();
+  }
+  FinishTask finish_task;
+  q.ProcessArchivedFiles([&total_size](bool is_file, const std::string & full_file_path) {
+    TEST_EQUAL(true, is_file);
+    const std::string data = FileManager::ReadFileAsString(full_file_path);
+    TEST_EQUAL(total_size, data.size());
+    // Integrity check.
+    size_t beg = 0, end = 0;
+    while (beg < data.size()) {
+      end += data[beg];
+      const size_t count = end - beg;
+      TEST_EQUAL(std::string(static_cast<size_t>(data[beg]), count), data.substr(beg, count));
+      beg += count;
+    }
+    total_size = 0;
+    return true;
+  }, std::bind(&FinishedCallback, std::placeholders::_1, std::ref(finish_task)));
+  TEST_EQUAL(ProcessingResult::EProcessedSuccessfully, finish_task.get());
+  TEST_EQUAL(size_t(0), total_size);  // Zero means that processor was called.
+}
+
+int main(int, char * []) {
+  Test_ScopedRemoveFile();
```

Comment:
If you ended up re-implementing `googletest`, at least make sure each test is invoked automatically.